src/main/java/org/alfresco/repo/site/SiteGroupCannedQuery.java

```diff
@@ -0,0 +1,105 @@
+ * (at your option) any later version.
+ *
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+package org.alfresco.repo.site;
+
+import org.alfresco.query.AbstractCannedQuery;
+import org.alfresco.query.CannedQueryParameters;
+import org.alfresco.query.CannedQuerySortDetails;
+import org.alfresco.query.CannedQuerySortDetails.SortOrder;
+import org.alfresco.service.cmr.security.AuthorityService;
+import org.alfresco.service.cmr.site.SiteInfo;
+import org.alfresco.service.cmr.site.SiteService;
+import org.alfresco.service.cmr.site.SiteService.SiteMembersCallback;
+import org.alfresco.util.Pair;
+
+import java.util.*;
+
+// TODO currently have to read all sites into memory for sorting purposes. Find a way that doesn't
+public class SiteGroupCannedQuery extends AbstractCannedQuery<SiteMembership>
+{
+  private AuthorityService authorityService;
+  private SiteService siteService;
+
+  protected SiteGroupCannedQuery(
+      SiteService siteService,
+      AuthorityService authorityService,
+      CannedQueryParameters parameters) {
+    super(parameters);
+    this.authorityService = authorityService;
+    this.siteService = siteService;
+  }
+
+  @Override
+  protected List<SiteMembership> queryAndFilter(CannedQueryParameters parameters) {
+    SiteMembersCannedQueryParams paramBean = (SiteMembersCannedQueryParams) parameters.getParameterBean();
+
+    String siteShortName = paramBean.getShortName();
+    CannedQuerySortDetails sortDetails = parameters.getSortDetails();
+
+    final CQSiteGroupsCallback callback = new CQSiteGroupsCallback(siteShortName, sortDetails.getSortPairs());
+    siteService.listMembers(siteShortName, null, null, false, true, paramBean.isExpandGroups(), callback);
+    callback.done();
+
+    return callback.getSiteMembers();
+  }
+
+  @Override
+  protected boolean isApplyPostQuerySorting() {
+    // already sorted as a side effect by CQSiteMembersCallback
+    return false;
+  }
+
+  private class CQSiteGroupsCallback implements SiteMembersCallback {
+    private SiteInfo siteInfo;
+    private Set<SiteMembership> siteGroups;
+
+    CQSiteGroupsCallback(String siteShortName, List<Pair<? extends Object, SortOrder>> sortPairs) {
+      this.siteInfo = siteService.getSite(siteShortName);
+      this.siteGroups = sortPairs != null && sortPairs.size() > 0
+              ? new TreeSet<>(new SiteMembershipComparator(sortPairs, SiteMembershipComparator.Type.MEMBERS))
+              : new HashSet<>();
+    }
+
+    @Override
+    public void siteMember(String authority, String role) {
+		if(authorityService.authorityExists(authority)) {
```

Comment:
Not sure, but isn't siteMember() invoked for users and for groups? So just checking if the authority exists won't filter out the groups

Reply:
This call back is controlled in the method param. new param added to get only the required call back. i mean 
**includeUsers**      includes the users
**includeGroups**    include the groups
  **expandGroups**      true if expand group member into user list, false otherwise